src/task_manager.cc

```diff
@@ -143,16 +154,89 @@ namespace detail {
 
 		if(auto cgid = tsk->get_collective_group_id(); cgid != 0) {
 			if(auto prev = last_collective_tasks.find(cgid); prev != last_collective_tasks.end()) {
-				tsk->add_dependency({task_map.at(prev->second).get(), dependency_kind::ORDER_DEP});
+				add_dependency(tsk.get(), task_map.at(prev->second).get(), dependency_kind::ORDER_DEP);
 				last_collective_tasks.erase(prev);
 			}
 			last_collective_tasks.emplace(cgid, tid);
 		}
 	}
 
-	void task_manager::invoke_callbacks(task_id tid) {
+	task_id task_manager::get_new_tid() { return next_task_id++; }
+
+	task& task_manager::register_task_internal(std::unique_ptr<task> task) {
+		auto& task_ref = *task;
+		assert(task != nullptr);
+		task_map.emplace(task->get_id(), std::move(task));
+		execution_front.insert(&task_ref);
+		return task_ref;
+	}
+
+	void task_manager::invoke_callbacks(task_id tid, task_type type) {
 		for(auto& cb : task_callbacks) {
-			cb(tid);
+			cb(tid, type);
+		}
+	}
+
+	void task_manager::add_dependency(task* depender, task* dependee, dependency_kind kind) {
+		assert(depender != dependee);
+		assert(depender != nullptr && dependee != nullptr);
+		depender->add_dependency({dependee, kind});
+		execution_front.erase(dependee);
+		max_pseudo_critical_path_length = std::max(max_pseudo_critical_path_length, depender->get_pseudo_critical_path_length());
+	}
+
+	bool task_manager::need_new_horizon() const { //
```

Comment:
What's with the empty comment? And this could probably also be moved into the header.

Reply:
the empty comment was to deal with the auto formatter doing weird formatting. but I will move it to into the header.